src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs

```diff
@@ -4923,6 +4959,58 @@ public async Task SendToAnotherClientFromOnConnectedAsync()
         }
     }
 
+    [Fact]
+    public async Task ConnectionClosesWhenClientSendsCloseMessage()
+    {
+        using (StartVerifiableLog())
+        {
+            var state = new ConnectionLifetimeState();
+            var serviceProvider = HubConnectionHandlerTestUtils.CreateServiceProvider(s => s.AddSingleton(state), LoggerFactory);
+            var connectionHandler = serviceProvider.GetService<HubConnectionHandler<ConnectionLifetimeHub>>();
+
+            using var client = new TestClient();
+
+            var connectionHandlerTask = await client.ConnectAsync(connectionHandler);
+
+            await client.SendHubMessageAsync(new CloseMessage(error: null));
+
+            var message = Assert.IsType<CloseMessage>(await client.ReadAsync().DefaultTimeout());
+            Assert.Null(message.Error);
+
+            await connectionHandlerTask.DefaultTimeout();
+
+            Assert.Null(state.DisconnectedException);
```

Comment:
Are we eventually going to ensure that the `OnDisconnectedAsync` exception is not null when the client supports the close message but the transport closed "gracefully" without a close message?

Reply:
I don't think we can rely on that hope. In the original issue the entire premise was that "In some cases (based on reports to the "ASP.NET Discussions" mailing list), Hub.OnDisconnectedAsync can be called without an Exception when SignalR clients disconnect ungracefully."

Reply:
I didn't mean that that's what happens today, but the end result of the current train of work is that there should be an exception for any un-graceful close.